### ISSUE TITLE
Llama 3 Preferred RAG Prompting Format (xml tags vs. markdown vs. something else)

### ISSUE THREAD

**Issue by krumeto** [meta-llama/llama-recipes#450]
### 🚀 The feature, motivation and pitch

Anthropic directly states that their models prefer context for longer prompts (like the usual RAG applications) to be inserted in XML tags. Some claim OpenAI's models prefer markdown-style (their docs mention both markdown and XML tags).

Does Llama 3 have a preferred format for longer prompts? 

Thank you in advance!

**Comment by jeffxtang** [comment]
There's no mention of a preferred format for Llama 3. According to the [Llama 3 model card prompt format](https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3/), you just need to follow the new Llama 3 format there (also specified in HF's blog [here](https://huggingface.co/blog/llama3#how-to-prompt-llama-3)), but if you use a framework LangChain or service provider like Groq/Replicate or run Llama 3 locally using Ollama for your RAG apps, most likely you won't need to deal with the new prompt format directly as it's been hardcoded by them under the hood. Just use an appropriate RAG prompt (e.g. [rag-prompt](https://smith.langchain.com/hub/rlm/rag-prompt)) with your question, context and possibly chat history for Llama 3 to answer.

**Comment by jeffxtang** [comment]
I'm not aware of such preference for Llama 3, but it should be easy, with some automated RAG evaluation frameworks (there're quite a few nice open source frameworks), to compare the results of the example with different formats and see if there's any quality difference.  @krumeto

**Comment by trivikramak** [comment]
Hi @krumeto, were you able to find out what works best with Llama-3 ?

**Comment by krumeto** [comment]
Hey @trivikramak, no, not yet (apologies).

**Comment by scottstirling** [comment]
Ask it and see what it says.  Try some different stuff.  Interesting question.

**Comment by mindful-time** [comment]
I using llama3:instruct and llama3:latest model from Ollama for this and seems like providing system prompt with chat history and context just doesn't work out of the box. I am getting replies from the both where it just keep giving me the entire system prompt as the answer when just asking Hi. Seems like it's not working like you expect with OpenAI , also using langchain's chatollama class for it. So this will definitely help.

**Comment by init27** [comment]
Hi @krumeto Thanks for the great Q! 

Please re-open the issue if you have any interesting new finds. 

In my tests llama 3.1 model(s) are agnostic to the prompt formatting if it's set in the system prompt.

**Comment by uninterestedcode** [comment]
So I realize that this is quite the old threat but I figured I'd share something I've found

An interesting way to include a document you is by sending a message with a custom header_id whereby your change the section where user or assistant is usually stated you instead put some custom role name ie. Document, list. I've found that llama 3 does understand this quite easily and I've seen anecdotally less errors/mistakes in output. I've used this for summarization, yaml json plain text conversions, extraction of info etc.

On another note I've also found that changing the rolename  of user and assistant can also be helpful, ie. Changing the generation's output header_id can help 'force' the model to output in a given format. It usually eliminates additional comment by the model such as: sure I'll summarize the chat etc. 

This is all anecdotal, so usage will vary, but I've found added benefit from doing this
